people_recognition_3d/src/people_recognition_3d/people_recognizer_3d.py

```diff
@@ -1,26 +1,52 @@
 #!/usr/bin/env python
-from __future__ import division
+from __future__ import print_function, division
 
 import math
 import PyKDL as kdl
 from collections import namedtuple
 from itertools import groupby
+import numpy as np
 
 import image_geometry
-import message_filters
-import numpy as np
 import rospy
 import tf
 from cv_bridge import CvBridge
 from geometry_msgs.msg import Point, Vector3, Pose, Quaternion
-from image_recognition_msgs.srv import Recognize
-from tue_msgs.msg import Person, People
 from sensor_msgs.msg import Image, CameraInfo
 from std_msgs.msg import ColorRGBA
 from visualization_msgs.msg import Marker, MarkerArray
 
-Joint = namedtuple('Joint', ['group_id', 'name', 'p', 'point'])
+from people_recognition_msgs.srv import RecognizePeople2D, RecognizePeople2DResponse
+from people_recognition_msgs.msg import Person3D
+
+def _get_and_wait_for_service(srv_name, srv_class):
+    """
+    Function to start and wait for dependent service
+    :param: srv_name: Service name
+    :param: srv_class: Service class
+    :return: started ServiceProxy object
+    """
+    service = rospy.ServiceProxy('{}'.format(srv_name), srv_class)
+    rospy.loginfo("Waiting for service {} ...".format(service.resolved_name))
+    service.wait_for_service()
+    return service
 
+def _get_service_response(srv, args):
+    """
+    Method to get service response with checks
+    :param: srv: service
+    :param: args: Input arguments of the service request
+    :return: response
+    """
+    response = None
+    try:
+        response = srv(args)
+    except rospy.ServiceException as e:
+        rospy.logwarn("{} service call failed: {}".format(srv.resolved_name, e))
+        return None
+    return response
```

Comment:
When you try-catch the exception, you will loose all information WHY a service call has failed. Please catch in the right spot, only when you can handle it.

Reply:
I don't understand your point here. Catching `rospy.ServiceException as e` and then logging it will show the reason why the ros service exception was raised.

Reply:
I'll try to explain it.

The function `_get_service_response` returns `None` or `RecognizeResponse`. When you receive `None`, you have no idea why the call failed. If you had let the exception bubble, you have a typed reason why the call failed. There can be multiple reasons: the service was not online, the arguments of the service call were wrong, processing took too long. With exceptions you can do the following:
```py
try:
	somecall(args)
except FileNotFoundError:
	# config file is not yet created, let's create it
except RuntimeError:
	# arguments to the function were wrong, let's give the user feedback in the GUI
except Exception:
	# I don't know what happened, let's abort the code for safety reasons
```
In your example, you immediately assert the resonse. So a `ServiceException` gets converted to a `AssertionError`. You loose information, that was not the reasons why the function failed.
```py
openpose_response = _get_service_response(self._openpose_srv, image_msg)
assert isinstance(openpose_response, RecognizeResponse)
```

Better (I think)
```py
openpose_response = _get_service_response(self._openpose_srv, image_msg)
# just let the exception bubble up, there is nothing you can do here
# the message handler will log the error the the console with a stack trace.
```

Reply:
Does this make it more clear?

Reply:
I see your point. IMO, this would be better:
```python
import sys

try:
    response = srv(args)
except:
    e = sys.exc_info()[0:2]
    rospy.logwarn("{}".format(e))
    return None

return response
```

And on the other side:
```python
openpose_response = _get_service_response(self._openpose_srv, image_msg)
if not openpose_response:
    raise rospy.ServiceException("Openpose failed")
```

This way all the exceptions will be caught and put in the log file. What do you think about this?

Reply:
That's better. If you have the same logic in the same place, you could also factor this out:
```py
def image_callback(msg):
	try:
		image_callback_wrapped(msg)
	except Exception as e:
		rospy.logerr("Exception while handling an image:\n", exc_info=True)  # I'm not sure if exc_info works, but a traceback is really handy

def image_callback_wrapped(msg):
	msg = someoperation(msg)
	data = srv(msg)  # just let it raise
	result = otheroperation(msg, data)
	rospy.Publish(result)
```

Reply:
Which is what the default image callback already does for you

Reply:
https://github.com/tue-robotics/people_recognition/issues/10 
Don't let perfect be the enemy of good.